### ISSUE TITLE
Gson adapter should support null values

### ISSUE THREAD

**Issue by almozavr** [immutables/immutables#127]
For now, any field like `ImmutableVideo video = null` will fail gson serializing process 'cause generated adapter doesn't support null value, and it looks like this: ![](http://monosnap.com/image/e9Wta6ZrlKJaOCwwkfKJOuyozBA7qT.png)

Maybe, some additional check needed?
```java
    @Override
    public void write(JsonWriter out, Video value) throws IOException {
      if (value == null) out.nullValue();
      else writeVideo(out, value);
    }
```

**Comment by elucash** [comment]
Hi Alexey
Thank you for the report. I'm in uncertainty whether this is by design or something that could should fix.

In theory, when there's `@Nullable` attributes and appropriate null handling code is generated.
Then, when you have a type adapter instance you can turn it into null safe by calling `adapter.nullSafe()` to get adapter instance which handles nulls gracefully.

I'm curious how you use it, from the stack trace on the picture I can figure out that there is pojo (reflective type adapter) which have Video as it's field. This field could be null in the model or in JSON.

Given that immutables have null checks on attribute initializers/constructors, we could implement the kind of check you proposed, to make type adapters null-safe by default. While it was not initially intended it might help make mixing null-allowing pojos and immutables easier.

**Comment by almozavr** [comment]
Thanks for a reply! Let's dig into it.

> In theory, when there's @Nullable attributes and appropriate null handling code is generated.
Then, when you have a type adapter instance you can turn it into null safe by calling adapter.nullSafe() to get adapter instance which handles nulls gracefully.
Didn't get it if you're thinking about API intent or this is something already implemented? Anyway, I haven't found `adapter.nullSafe()` and not sure why it's needed, even more I would vote for default `nullSafe` support, see why.

`Gson` gives total control of  serialization/deserialization to `TypeAdapter`, so even null-values could be serialized as `{video="I'm null, haha"}` - it's up to `TypeAdapter` by design. So, when generated `immutable-type` adapter receives null to write, it *has* to provide something to `JsonWriter` (by default it should be `nullValue`, so `Gson` and `JsonWriter` could process it [gracefully](https://sites.google.com/site/gson/gson-user-guide#TOC-Null-Object-Support).

P.S. The case we are talking about is:
```java
public class Container { // this class we're serializing
  ImmutableVideo video // or Video video
}
...
@Value.Immutable
@Gson.TypeAdapters
public abstract class Video {
...
}
```

**Comment by almozavr** [comment]
As a workaround here is a wrapper to check null-safety and to prevent `NullPointerException`:
```java
    public static class NullSafeImmutableAdapterFactory implements TypeAdapterFactory {

        private final TypeAdapterFactory factory;

        public NullSafeImmutableAdapterFactory(TypeAdapterFactory factory) {
            this.factory = factory;
        }

        @Override public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
            TypeAdapter<T> adapter = factory.create(gson, type);
            if (adapter == null) return null;
            //
            return new TypeAdapter<T>() {
                @Override public void write(JsonWriter out, T value) throws IOException {
                    if (value == null) out.nullValue();
                    else adapter.write(out, value);
                }

                @Override public T read(JsonReader in) throws IOException {
                    if (in.hasNext()) return adapter.read(in);
                    else return null;
                }
            };
        }
    }
```

**Comment by elucash** [comment]
At least I found this in `TypeAdapter`'s javadoc:
```
... In general, a type adapter is required to handle nulls in write and read methods. ...
```
I guess we just have to add graceful null handling in generated type adapters. As immutables are null-hostile by default, non-nullable attributes of value objects will not accept null values even if type adapter was able to deserialize null from JSON, however `@Nullable` attributes will work with JSON nulls

Just to make a note, workaround that you described seems not completely correct at reading nulls, so If you ever need such wrapper `nullSafe()` method could be used (see implementation)

```java
public static TypeAdapterFactory asNullSafe(final TypeAdapterFactory factory) {
  return new TypeAdapterFactory() {
    @Override public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
      TypeAdapter<T> adapter = factory.create(gson, type);
      return adapter != null
          ? adapter.nullSafe()
          : null;
    }
  };
}
```

**Comment by elucash** [comment]
The fix was released in 2.0.15. Now generated type adapters handle nulls as built-in type adapters.
Thanks!

**Comment by almozavr** [comment]
@elucash Thanks for a lightning-fast fix, it's  :+1:
